File: server/src/services/dbBackup.ts
import fs from 'fs';
import path from 'path';
import { config } from '../config.js';
import { logger } from '../utils/logger.js';

const BACKUP_DIR = path.join(config.dataDir, 'backups');
const MAX_BACKUPS = 7; // Keep 7 daily backups

export function backupDatabase(): void {
  try {
    // Create backup directory if it doesn't exist
    if (!fs.existsSync(BACKUP_DIR)) {
      fs.mkdirSync(BACKUP_DIR, { recursive: true });
    }

    // Generate backup filename with current date (YYYY-MM-DD)
    const timestamp = new Date().toISOString().split('T')[0];
    const backupPath = path.join(BACKUP_DIR, `backup-server-${timestamp}.db`);
    const dbPath = path.join(config.dataDir, 'backup-server.db');

    // Copy database file if it exists
    if (fs.existsSync(dbPath)) {
      fs.copyFileSync(dbPath, backupPath);
      logger.info({ backupPath }, '[DB BACKUP] Created database backup');

      // Clean up old backups
      cleanOldBackups();
    } else {
      logger.warn('[DB BACKUP] Database file not found, skipping backup');
    }
  } catch (error) {
    logger.error({ error }, '[DB BACKUP] Failed to create backup');
  }
}

function cleanOldBackups(): void {
  try {
    const files = fs.readdirSync(BACKUP_DIR)
      .filter(f => f.startsWith('backup-server-') && f.endsWith('.db'))
      .map(f => ({
        name: f,
        path: path.join(BACKUP_DIR, f),
        mtime: fs.statSync(path.join(BACKUP_DIR, f)).mtime
      }))
      .sort((a, b) => b.mtime.getTime() - a.mtime.getTime());

    // Delete backups beyond MAX_BACKUPS
    const toDelete = files.slice(MAX_BACKUPS);
    toDelete.forEach(f => {
      fs.unlinkSync(f.path);
      logger.info({ backup: f.name }, '[DB BACKUP] Deleted old backup');
    });

    if (toDelete.length > 0) {
      logger.info({ kept: MAX_BACKUPS, deleted: toDelete.length }, '[DB BACKUP] Cleanup completed');
    }
  } catch (error) {
    logger.error({ error }, '[DB BACKUP] Failed to clean old backups');
  }
}

File: backup-server-rs/src/services/db_backup.rs
use std::path::Path;

const MAX_BACKUPS: usize = 7;

pub fn backup_database(db_path: &str, data_dir: &Path) -> anyhow::Result<()> {
    let backup_dir = data_dir.join("backups");
    std::fs::create_dir_all(&backup_dir)?;

    let today = chrono::Local::now().format("%Y-%m-%d").to_string();
    let backup_name = format!("backup-server-{}.db", today);
    let backup_path = backup_dir.join(&backup_name);

    if backup_path.exists() {
        tracing::info!("[DB Backup] Today's backup already exists, skipping");
        return Ok(());
    }

    std::fs::copy(db_path, &backup_path)?;
    tracing::info!("[DB Backup] Created backup: {}", backup_name);

    // Cleanup old backups
    let mut backups: Vec<_> = std::fs::read_dir(&backup_dir)?
        .filter_map(|e| e.ok())
        .filter(|e| {
            e.file_name()
                .to_string_lossy()
                .starts_with("backup-server-")
        })
        .collect();

    backups.sort_by(|a, b| b.file_name().cmp(&a.file_name()));

    for old in backups.into_iter().skip(MAX_BACKUPS) {
        let _ = std::fs::remove_file(old.path());
        tracing::info!("[DB Backup] Removed old backup: {}", old.file_name().to_string_lossy());
    }

    Ok(())
}
